File: src/Day4/part1.go
package Day4

import (
  "strconv"
  "strings"
)

func Day4Part1(input string) (string, error) {
  lines := strings.Split(strings.TrimSpace(input), "\n\n")
  passports := parsePassports(lines)
  validPassports := make([]Passport, 0)
  for _, passport := range passports {
    if v := isValid1(passport); v {
      validPassports = append(validPassports, passport)
    }
  }
  return strconv.Itoa(len(validPassports)), nil
}

func isValid1(p Passport) bool {
  return p.Byr != "" &&
    p.Iyr != "" &&
    p.Eyr != "" &&
    p.Hgt != "" &&
    p.Hcl != "" &&
    p.Ecl != "" &&
    p.Pid != ""
}

File: src/Day4/part1.rs
use crate::day4::util::{Passport, parse_passports};

pub fn day4part1(input: &String) -> Option<String> {
    let passports = parse_passports(&input);
    let valid_passports: Vec<&Passport> = passports
        .iter()
        .filter(|p| is_valid(p))
        .collect();
    return Some(valid_passports.len().to_string());
}

fn is_valid(passport: &Passport) -> bool {
    return passport.byr != "" &&
        passport.iyr != "" &&
        passport.eyr != "" &&
        passport.hgt != "" &&
        passport.hcl != "" &&
        passport.ecl != "" &&
        passport.pid != "";
}
